src/service_address.rs

```diff
@@ -1,160 +1,39 @@
-                device_id.into(),
-            ),
-        }
-    }
-
-    #[deprecated]
-    pub fn new_aci(uuid: Uuid) -> Self {
-        Self::from_aci(uuid)
-    }
-
-    pub fn from_aci(uuid: Uuid) -> Self {
-        Self {
-            uuid,
-            identity: ServiceIdType::AccountIdentity,
-        }
-    }
-
-    #[deprecated]
-    pub fn new_pni(uuid: Uuid) -> Self {
-        Self::from_pni(uuid)
-    }
-
-    pub fn from_pni(uuid: Uuid) -> Self {
-        Self {
-            uuid,
-            identity: ServiceIdType::PhoneNumberIdentity,
-        }
-    }
-
-    pub fn aci(&self) -> Option<libsignal_protocol::Aci> {
-        use libsignal_protocol::Aci;
-        match self.identity {
-            ServiceIdType::AccountIdentity => {
-                Some(Aci::from_uuid_bytes(self.uuid.into_bytes()))
-            },
-            ServiceIdType::PhoneNumberIdentity => None,
-        }
-    }
-
-    pub fn pni(&self) -> Option<libsignal_protocol::Pni> {
-        use libsignal_protocol::Pni;
-        match self.identity {
-            ServiceIdType::AccountIdentity => None,
-            ServiceIdType::PhoneNumberIdentity => {
-                Some(Pni::from_uuid_bytes(self.uuid.into_bytes()))
-            },
-        }
-    }
-
-    pub fn to_service_id(&self) -> String {
-        self.to_string()
-    }
-}
-
-impl From<ServiceId> for ServiceAddress {
-    fn from(service_id: ServiceId) -> Self {
-        match service_id {
-            ServiceId::Aci(service_id) => {
-                ServiceAddress::from_aci(service_id.into())
-            },
-            ServiceId::Pni(service_id) => {
-                ServiceAddress::from_pni(service_id.into())
-            },
-        }
-    }
-}
-
-impl TryFrom<&ProtocolAddress> for ServiceAddress {
-    type Error = ParseServiceAddressError;
-
-    fn try_from(addr: &ProtocolAddress) -> Result<Self, Self::Error> {
-        let value = addr.name();
-        if let Some(pni) = value.strip_prefix("PNI:") {
-            Ok(ServiceAddress::from_pni(Uuid::parse_str(pni)?))
-        } else {
-            Ok(ServiceAddress::from_aci(Uuid::parse_str(value)?))
-        }
-        .map_err(|e| {
-            tracing::error!("Parsing ServiceAddress from {:?}", addr);
-            ParseServiceAddressError::InvalidUuid(e)
-        })
+        let service_id: ServiceId = self.into();
+        ProtocolAddress::new(service_id.service_id_string(), device_id.into())
     }
-}
-
-impl TryFrom<&str> for ServiceAddress {
-    type Error = ParseServiceAddressError;
 
-    fn try_from(value: &str) -> Result<Self, Self::Error> {
-        if let Some(pni) = value.strip_prefix("PNI:") {
-            Ok(ServiceAddress::from_pni(Uuid::parse_str(pni)?))
-        } else {
-            Ok(ServiceAddress::from_aci(Uuid::parse_str(value)?))
+    fn aci(self) -> Option<Aci> {
+        match self.into() {
+            ServiceId::Aci(aci) => Some(aci),
+            ServiceId::Pni(_) => None,
```

Comment:
We can maybe consider upstreaming these two convenience methods.

Reply:
https://github.com/signalapp/libsignal/pull/595

Reply:
Cool, thanks!